crates/astria-conductor/src/executor/client.rs

```diff
@@ -168,3 +216,172 @@ fn block_identifier(number: u32) -> raw::BlockIdentifier {
+            "failed executing RPC; retrying after after backoff"
+        );
+        futures::future::ready(())
+    }
+}
+
+fn retry_config()
+-> tryhard::RetryFutureConfig<tryhard::backoff_strategies::ExponentialBackoff, OnRetry> {
+    tryhard::RetryFutureConfig::new(u32::MAX)
+        .exponential_backoff(Duration::from_millis(100))
+        // XXX: This should probably be configurable.
+        .max_delay(Duration::from_secs(10))
+        .on_retry(OnRetry {
+            parent: Span::current(),
+        })
+}
+
+/// An exponential retry strategy branching [`tonic::Status::code`].
+///
+/// Execution will be retried under the following conditions:
+///
+/// ```text
+/// Code::Cancelled
+/// Code::Unknown
+/// Code::DeadlineExceeded
+/// Code::NotFound
+/// Code::ResourceExhausted
+/// Code::Aborted
+/// Code::Unavailable
+/// Code::DataLoss
+/// ```
+struct ExecutionApiRetryStrategy {
+    delay: Duration,
+}
+
+impl<'a> BackoffStrategy<'a, tonic::Status> for ExecutionApiRetryStrategy {
+    type Output = RetryPolicy;
+
+    fn delay(&mut self, _attempt: u32, error: &'a tonic::Status) -> Self::Output {
+        if should_retry(error) {
+            let prev_delay = self.delay;
+            self.delay = self.delay.saturating_mul(2);
+            RetryPolicy::Delay(prev_delay)
+        } else {
+            RetryPolicy::Break
+        }
+    }
+}
+
+fn should_retry(status: &tonic::Status) -> bool {
+    use tonic::Code;
+    // gRPC return codes and if they should be retried. Also refer to
+    // [1] https://github.com/grpc/grpc/blob/1309eb283c3e11c471191f286ceab01b75477ffc/doc/statuscodes.md
+    //
+    // Code::Ok => no, success
+    // Code::Cancelled => yes, but should be revisited if "we" would cancel
+    // Code::Unknown => yes, could this be returned if the endpoint is unavailable?
+    // Code::InvalidArgument => no, no point retrying
+    // Code::DeadlineExceeded => yes, server might be slow
+    // Code::NotFound => yes, resource might not yet be available
+    // Code::AlreadyExists => no, no point retrying
+    // Code::PermissionDenied => no, execution API uses permission-denied restart-trigger
+    // Code::ResourceExhausted => yes, retry after a while
+    // Code::FailedPrecondition => no, failed precondition should not be retried unless the
+    //                             precondition is fixed, see [1]
+    // Code::Aborted => yes, although this applies to a read-modify-write sequence. We should
+    //                  implement this not per-request but per-request-sequence (for example,
+    //                  execute + update-commitment-state)
+    // Code::OutOfRange => no, we don't expect to send any out-of-range requests.
+    // Code::Unimplemented => no, no point retrying
+    // Code::Internal => no, this is a serious error on the backend; don't retry
+    // Code::Unavailable => yes, retry after backoff is desired
+    // Code::DataLoss => yes, unclear how this would happen; err on retrying
```

Comment:
I initially thought this meant only retry once and return an err when retrying if we get the same error again! :D  Maybe it could be phrased "err on the side of retrying"?

Reply:
Switched it to `=> no` in  40a9f51 :)